### ISSUE TITLE
Word embeddings for string representation.

### ISSUE THREAD

**Issue by Manthan-R-Sheth** [mlpack/mlpack#1276]
This is a prototype of the numeric representation of strings.
```
embed = model.Embedding(Vocab_size, Dimen_size) // Embedding Layer
embed.weight = (pretrained_weight) // matrix datatype
```
`pretrained_weight` can be word vector, GLoVE or any other already trained dense representation. 
Also each word is indexed by the id, which allows for word embedding lookup.

Steps that I think would be needed :-
1. A parser has to be written/modified to load the csv for pretrained words. Once a `arma::mat` is formed, it can be used for all algorithms. (CLI, cpp programs as well as bindings).
2. Support for different dimensions representations.

@zoq @rcurtin  if you could give your inputs related to this, it would be very helpful.

**Comment by rcurtin** [comment]
Can you clarify what you mean by 'each word is indexed by the id'?  I guess the most important thing (to me) is the representation we use in C++ for a string dataset---unless we instead convert directly and immediately to a numeric representation.  I think that once we have a clear picture of what to do there, the rest of the pieces will easily fall into place.

**Comment by Manthan-R-Sheth** [comment]
What I mean by 'each word indexed by the id' is that each word in the input will have a unique id associated to it. This will be done by the `tokeniser`(this will be implemented) that identifies the unique words in the input samples. Suppose our vocab size is 50 words (which we find after tokeniser does first pass through the data). Each word will also be mapped to a unique id. For eg:- The sentence "This is the name of the first commit" will be converted to [ 17,12,45,32,43,45,27,29 ] after the tokeniser is executed.
This is the input for the Embedding Layer. Thus input is considered to be a matrix of word indices.
This associated id will be the row number in the embedding matrix that is learnt for the words in the vocab or used directly by pre-trained vectors.

**Comment by Manthan-R-Sheth** [comment]
@rcurtin 
Any views on the above idea?

**Comment by scrcro** [comment]
I am currently working on an Embedding layer. You can write your own layers now with the latest drop, therefore you will have control over the `Forward`, `Backward` (not so important as this will be the first layer usually) and `Gradient` function. There are a couple of issues I am facing too:
- The optimization right now will update all your weight vectors and not just the relevant ones. This could be a problem for very large vocabularies
- If you want to use RNN then it expects one output for every timestep. This is a bit of a problem for n-gram models. But if you stick to FFN you should be ok.

Also using a pre-trained vector may not be good idea if you are not solving a problem in the same domain. The idea is to train the weights of the embedded layer too.

**Comment by Manthan-R-Sheth** [comment]
@scrcro I agree with you. Even I faced the same issues. We would have to think of a clever solution to update only required rows in the embedding matrix.
Usually pre-trained vectors like word2vec or GLoVE and more recently CLoVE and fast text, have seen to be used across variety of tasks(so flexibility to use pre-trained vectors would be better). Also, beginning with these vectors and then fine tuning them during training (transfer learning) is widely used (which essentially becomes similar to training embedding matrix also).

**Comment by scrcro** [comment]
If you write a custom layer, you can do the trick in your layer. Do not pass the actual weights as the Parameters() return value. Return a small dummy and do the actual weight updates in the `Gradient` method of your custom layer. This of course is assuming this layer is the first layer.
You have to ensure that your layer uses the same optimizer as the other layers in the network.

**Comment by Manthan-R-Sheth** [comment]
Please tell me what i understood is correct or not.
In the `Gradient` method, we have the unique word index of the input and so we can update the corresponding row of the embedding matrix. This way we need not update the entire matrix, it brings about just row updates right?

**Comment by rcurtin** [comment]
@Manthan-R-Sheth: please be patient; I am sure you can see the number of requests we must respond to.  I answer when I am able to make the time.

> What I mean by 'each word indexed by the id' is that each word in the input will have a unique id associated to it.

Got it, that makes sense.  The things I would encourage you to think about are:

 * What will be the representation of a dataset containing strings in C++?  Will you convert to `arma::mat` (or similar) at load time or later?

 * Will there be a clear way to return a word given an index inside of a program that a user writes?

**Comment by Manthan-R-Sheth** [comment]
@rcurtin 
Thanks for the response, I understand the heavy workload you are undergoing. I am figuring out a way for the points you mentioned already. The `tokeniser` class that I will implement will contain the functionalities related to the mapping.

**Comment by scrcro** [comment]
@Manthan-R-Sheth 
The `Gradient` method takes the input that was provided to the network, and the error propagated from the next layer and calculates the gradient for this layer. So yes you can update your weights here. Have a look at `LinearNoBias` class.

**Comment by Manthan-R-Sheth** [comment]
@scrcro 
Thats true. Infact I had submitted a PR already on similar lines. you can find it at #1277

**Comment by scrcro** [comment]
Good start. Though the real trick is in updating the weights. Thoughts on how you want to attempt this?

**Comment by Manthan-R-Sheth** [comment]
In the `Gradient` function since we have the input (which is the unique word indices), we know which row to update This row number is actually the word index itself. I think this should prevent the update of entire matrix. Only corresponding rows will be updated.

**Comment by scrcro** [comment]
Sounds good. Are you already working on this? Is this something that you can share/post?

**Comment by Manthan-R-Sheth** [comment]
I am in the planning phase for this right now. However, the direction seems correct. 
I had made the PR but as @rcurtin  rightly said, it was one of GSoC ideas, and so working on it now could lead to various code maintenance issues, the maximum I can do is share a rough prototype of the API and the idea altogether, once I plan it completely.
Sounds fine? @scrcro @rcurtin

**Comment by scrcro** [comment]
Sounds good. Will wait for it.

**Comment by rcurtin** [comment]
Closing this one for inactivity.  We can reopen if needed.
